src/serve.rs

```diff
@@ -119,12 +114,46 @@ impl ServeSystem {
         ));
         let router = router(state, cfg.clone());
         let addr = (cfg.address, cfg.port).into();
-        let server = Server::bind(&addr)
-            .serve(router.into_make_service())
-            .with_graceful_shutdown(shutdown_fut);
+        let crt = cfg.crt.as_os_str();
+        let key = cfg.key.as_os_str();
+        if !crt.is_empty() && !key.is_empty() {
+            Ok(tokio::spawn(async move {
+                if let Ok(config) =
+                    RustlsConfig::from_pem_file(cfg.crt.as_path(), cfg.key.as_path()).await
+                {
+                    tracing::info!("{} server listening at https://{}", SERVER, addr);
+                    let server =
+                        axum_server::bind_rustls(addr, config).serve(router.into_make_service());
+                    if let Err(err) = server.await {
+                        tracing::error!(error = ?err, "error from server task");
+                    }
+                } else {
+                    Self::start_http(shutdown_rx, addr, router).expect("http server start error.");
+                }
+            }))
+        } else {
+            Self::start_http(shutdown_rx, addr, router)
+        }
+    }
+
+    #[tracing::instrument(level = "trace", skip(shutdown_rx, addr, router))]
+    fn start_http(
+        mut shutdown_rx: broadcast::Receiver<()>,
+        addr: SocketAddr,
+        router: Router,
+    ) -> Result<JoinHandle<()>> {
+        // Build a shutdown signal for the warp server.
+        let shutdown_fut = async move {
+            // Any event on this channel, even a drop, should trigger shutdown.
+            let _res = shutdown_rx.recv().await;
+            tracing::debug!("server is shutting down");
+        };
 
         // Block this routine on the server's completion.
         tracing::info!("{} server listening at http://{}", SERVER, addr);
+        let server = Server::bind(&addr)
```

Comment:
Should this come before the tracing log?

Reply:
Changed comment position just before server.await.